packages/grafana-ui/src/components/Table/DefaultCell.tsx

```diff
@@ -51,51 +52,44 @@ export const DefaultCell: FC<TableCellProps> = (props) => {
+    }
+
+    if (field.config.custom?.displayMode === TableCellDisplayMode.ColorBackground) {
+      return { type: TableCellDisplayMode.ColorBackground, mode: TableCellBackgroundDisplayMode.Gradient };
+    }
+
+    if (field.config.custom?.displayMode === TableCellDisplayMode.ColorBackgroundSolid) {
+      return { type: TableCellDisplayMode.ColorBackground, mode: TableCellBackgroundDisplayMode.Basic };
+    }
+  }
+
+  return (field.config.custom as TableFieldOptions).cellOptions;
+}
+
 function getCellStyle(
   tableStyles: TableStyles,
-  field: Field,
+  cellOptions: TableCellOptions,
   displayValue: DisplayValue,
   disableOverflowOnHover = false
 ) {
   // How much to darken elements depends upon if we're in dark mode
   const darkeningFactor = tableStyles.theme.isDark ? 1 : -0.7;
 
-  // See if we're using deprecated settings
-  const usingDeprecatedSettings = field.config.custom?.displayMode !== undefined;
-
   // Setup color variables
   let textColor: string | undefined = undefined;
   let bgColor: string | undefined = undefined;
 
-  // Set colors using deprecated settings format
-  if (usingDeprecatedSettings) {
-    if (field.config.custom?.displayMode === TableCellDisplayMode.ColorText) {
-      textColor = displayValue.color;
-    } else if (field.config.custom?.displayMode === TableCellDisplayMode.ColorBackground) {
-      textColor = getTextColorForAlphaBackground(displayValue.color!, tableStyles.theme.isDark);
-      bgColor = tinycolor(displayValue.color).toRgbString();
-    } else if (
-      field.config.custom?.displayMode === TableCellDisplayMode.ColorBackground &&
-      field.config.custom?.backgroundDisplayMode === TableCellBackgroundDisplayMode.Gradient
-    ) {
-      const bgColor2 = tinycolor(displayValue.color)
-        .darken(10 * darkeningFactor)
-        .spin(5);
-      textColor = getTextColorForAlphaBackground(displayValue.color!, tableStyles.theme.isDark);
-      bgColor = `linear-gradient(120deg, ${bgColor2.toRgbString()}, ${displayValue.color})`;
-    }
-  }
-  // Set colors using updated sub-options format
-  else {
-    const cellDisplayMode = field.config.custom?.cellOptions?.mode;
-    const cellDisplayType = field.config.custom?.cellOptions?.type;
-
-    if (cellDisplayType === TableCellDisplayMode.ColorText) {
-      textColor = displayValue.color;
-    } else if (cellDisplayMode === TableCellBackgroundDisplayMode.Basic) {
+  if (cellOptions.type === TableCellDisplayMode.ColorText) {
+    textColor = displayValue.color;
+  } else if (cellOptions.type === TableCellDisplayMode.ColorBackground) {
+    if (cellOptions.mode === TableCellBackgroundDisplayMode.Basic) {
       textColor = getTextColorForAlphaBackground(displayValue.color!, tableStyles.theme.isDark);
       bgColor = tinycolor(displayValue.color).toRgbString();
-    } else if (cellDisplayMode === TableCellBackgroundDisplayMode.Gradient) {
+    } else if (cellOptions.mode === TableCellBackgroundDisplayMode.Gradient) {
```

Comment:
here you can see an example of Typescript union+descriminator typing working, cellOptions.mode is here fully typed automatically because it's inside the if statement of cellOptions.type === TableCellDisplayMode.ColorBackground